File: typescript-sdk/src/api/other.ts
/**
 * OtherApi — utility endpoints for services, plans, cost estimates, extras, and coupons.
 */

import { HttpClient } from "../http-client";
import { ApiResponse } from "../models/response";

export interface CostEstimateRequest {
  propertyId: number;
  planId: number;
  hours: number;
  couponCode?: string;
  extras?: number[];
}

export interface AvailableCleanersRequest {
  propertyId: number;
  date: string;  // YYYY-MM-DD
  time: string;  // HH:mm
}

export class OtherApi {
  constructor(private readonly http: HttpClient) {}

  /**
   * Return all available cleaning service types (e.g. Residential, Commercial, Airbnb).
   */
  getServices(): Promise<ApiResponse<unknown>> {
    return this.http.get("/v1/services");
  }

  /**
   * Return available booking plans for a given property.
   * @param propertyId  The property ID.
   */
  getPlans(propertyId: number): Promise<ApiResponse<unknown>> {
    return this.http.get("/v1/plans", { propertyId });
  }

  /**
   * Get the system-recommended cleaning hours based on property size.
   * Use this to pre-fill the hours field when creating a booking.
   * @param propertyId     The property ID.
   * @param bathroomCount  Number of bathrooms.
   * @param roomCount      Number of rooms/bedrooms.
   */
  getRecommendedHours(propertyId: number, bathroomCount: number, roomCount: number): Promise<ApiResponse<unknown>> {
    return this.http.get("/v1/recommended-hours", { propertyId, bathroomCount, roomCount });
  }

  /**
   * Calculate the estimated cost for a potential booking.
   * Use this to show a price preview before the user commits.
   * @param request  propertyId, planId, hours, and optional couponCode / extras.
   */
  getCostEstimate(request: CostEstimateRequest): Promise<ApiResponse<unknown>> {
    return this.http.post("/v1/cost-estimate", request);
  }

  /**
   * Get available add-on services for a given service type.
   * @param serviceId  The service type ID (from getServices).
   */
  getCleaningExtras(serviceId: number): Promise<ApiResponse<unknown>> {
    return this.http.get(`/v1/cleaning-extras/${serviceId}`);
  }

  /**
   * Find cleaners available for a specific property, date, and time slot.
   * @param request  propertyId, date (YYYY-MM-DD), and time (HH:mm).
   */
  getAvailableCleaners(request: AvailableCleanersRequest): Promise<ApiResponse<unknown>> {
    return this.http.post("/v1/available-cleaners", request);
  }

  /**
   * Return all valid coupon codes available for use at booking creation.
   */
  getCoupons(): Promise<ApiResponse<unknown>> {
    return this.http.get("/v1/coupons");
  }

  /**
   * List all cleaners, with optional status and search filters.
   * @param status  Filter by cleaner status ('active', 'inactive', 'pending').
   * @param search  Partial match against cleaner name or email.
   */
  listCleaners(status?: string, search?: string): Promise<ApiResponse<unknown>> {
    return this.http.get("/v1/cleaners", { status, search });
  }

  /**
   * Retrieve a single cleaner by their ID.
   * @param cleanerId  The cleaner's unique ID.
   */
  getCleaner(cleanerId: number): Promise<ApiResponse<unknown>> {
    return this.http.get(`/v1/cleaners/${cleanerId}`);
  }
}

File: python-sdk/cleanster/api/other.py
"""Other / Utilities API — services, plans, cost, extras, cleaners, coupons."""

from typing import Any, Dict, Optional

from ..http_client import HttpClient
from ..models.response import ApiResponse


class OtherApi:
    """
    Utility endpoints for reference data used in booking flows:
    services, plans, recommended hours, cost estimates, cleaning extras,
    available cleaners, and coupons.
    """

    def __init__(self, http: HttpClient):
        self._http = http

    def get_services(self) -> ApiResponse:
        """Return all available cleaning service types (e.g. Residential, Airbnb)."""
        raw = self._http.get("/v1/services")
        return ApiResponse.from_dict(raw)

    def get_plans(self, property_id: int) -> ApiResponse:
        """
        Return available booking plans for a given property.

        Args:
            property_id: The property ID.

        Returns:
            ApiResponse with data as a list of plan dicts.
        """
        raw = self._http.get("/v1/plans", params={"propertyId": property_id})
        return ApiResponse.from_dict(raw)

    def get_recommended_hours(
        self, property_id: int, bathroom_count: int, room_count: int
    ) -> ApiResponse:
        """
        Get the system-recommended cleaning hours based on property size.

        Args:
            property_id:     The property ID.
            bathroom_count:  Number of bathrooms.
            room_count:      Number of rooms/bedrooms.

        Returns:
            ApiResponse with data containing the recommended hours.
        """
        raw = self._http.get(
            "/v1/recommended-hours",
            params={
                "propertyId": property_id,
                "bathroomCount": bathroom_count,
                "roomCount": room_count,
            },
        )
        return ApiResponse.from_dict(raw)

    def get_cost_estimate(self, request: Dict[str, Any]) -> ApiResponse:
        """
        Calculate the estimated cost for a potential booking.

        Args:
            request: Dict with keys: propertyId, planId, hours,
                     and optionally couponCode, extras, etc.

        Returns:
            ApiResponse with data as a cost breakdown dict.
        """
        raw = self._http.post("/v1/cost-estimate", body=request)
        return ApiResponse.from_dict(raw)

    def get_cleaning_extras(self, service_id: int) -> ApiResponse:
        """
        Get available add-on services for a given service type.

        Args:
            service_id: The service type ID.

        Returns:
            ApiResponse with data as a list of extra dicts.
        """
        raw = self._http.get(f"/v1/cleaning-extras/{service_id}")
        return ApiResponse.from_dict(raw)

    def get_available_cleaners(self, request: Dict[str, Any]) -> ApiResponse:
        """
        Find cleaners available for a specific property, date, and time.

        Args:
            request: Dict with keys: propertyId, date, time.

        Returns:
            ApiResponse with data as a list of available cleaner dicts.
        """
        raw = self._http.post("/v1/available-cleaners", body=request)
        return ApiResponse.from_dict(raw)

    def get_coupons(self) -> ApiResponse:
        """Return all valid coupon codes available for use."""
        raw = self._http.get("/v1/coupons")
        return ApiResponse.from_dict(raw)
